File: app/src/test/java/com/matchthree/game/engine/BoardSimulationTest.kt
package com.matchthree.game.engine

import com.matchthree.game.model.BoardConfig
import com.matchthree.game.model.Position
import com.matchthree.game.rng.SeededRandom
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Simulates many generated 9x9/6 boards and reports the two tuning metrics named
 * in DESIGN.md: average legal-move count, and the probability that a random
 * adjacent swap is legal. Values print to the test stdout (both via the Fail
 * summary and standard stream) so they can be read from `./gradlew test`.
 */
class BoardSimulationTest {

    private val config = BoardConfig() // 9x9, 6 gem types

    @Test
    fun reportSimulationMetrics() {
        val boardCount = 150
        var totalLegalMoves = 0L
        var totalAdjacentPairs = 0L
        var totalLegalSwaps = 0L

        repeat(boardCount) { i ->
            val engine = GameEngine(config, SeededRandom(1000L + i))
            val board = engine.newGame()
            val legalCount = LegalMoveDetector.legalMoveCount(board)
            totalLegalMoves += legalCount

            for (row in 0 until config.height) {
                for (col in 0 until config.width) {
                    val a = Position(row, col)
                    if (board.gemAt(a) == null) continue
                    for (b in neighbours(row, col)) {
                        if (board.isInside(b) && board.gemAt(b) != null) {
                            totalAdjacentPairs++
                            if (engine.isLegalSwap(board, a, b)) totalLegalSwaps++
                        }
                    }
                }
            }
        }

        val avgLegalMoves = totalLegalMoves.toDouble() / boardCount
        val swapMatchProbability = totalLegalSwaps.toDouble() / totalAdjacentPairs

        println(
            "SIMULATION 9x9/6 over $boardCount boards -> " +
                "avg legal moves per board: $avgLegalMoves, " +
                "random-swap match probability: $swapMatchProbability",
        )

        // Loose sanity bounds so the test is a real gate but not flaky.
        assertTrue(avgLegalMoves > 1.0)
        assertTrue(avgLegalMoves < 80.0)
        assertTrue(swapMatchProbability > 0.0)
        assertTrue(swapMatchProbability < 0.5)
    }

    private fun neighbours(row: Int, col: Int): List<Position> = listOf(
        Position(row, col + 1),
        Position(row + 1, col),
    )
}

File: Tests/BoardSimulationTest.cs
using MatchThree.Engine.Model;
using MatchThree.Engine.Rules;
using NUnit.Framework;

namespace MatchThree.Engine.Tests;

public class BoardSimulationTest
{
    private readonly BoardConfig config = new();

    [Test]
    public void ReportSimulationMetrics()
    {
        const int boardCount = 150;
        long totalLegalMoves = 0;
        long totalAdjacentPairs = 0;
        long totalLegalSwaps = 0;

        for (var i = 0; i < boardCount; i++)
        {
            var engine = new GameEngine(config, new SeededRandom(1000L + i));
            var board = engine.NewGame();
            var legalCount = LegalMoveDetector.LegalMoveCount(board);
            totalLegalMoves += legalCount;

            for (var row = 0; row < config.Height; row++)
            {
                for (var col = 0; col < config.Width; col++)
                {
                    var a = new Position(row, col);
                    if (board.GemAt(a) is null) continue;
                    foreach (var b in Neighbours(row, col))
                    {
                        if (board.IsInside(b) && board.GemAt(b) is not null)
                        {
                            totalAdjacentPairs++;
                            if (engine.IsLegalSwap(board, a, b)) totalLegalSwaps++;
                        }
                    }
                }
            }
        }

        var avgLegalMoves = (double)totalLegalMoves / boardCount;
        var swapMatchProbability = (double)totalLegalSwaps / totalAdjacentPairs;

        TestContext.Progress.WriteLine(
            $"SIMULATION 9x9/6 over {boardCount} boards -> " +
            $"avg legal moves per board: {avgLegalMoves:F4}, " +
            $"random-swap match probability: {swapMatchProbability:F4}");
        TestContext.WriteLine(
            $"SIMULATION 9x9/6 over {boardCount} boards -> " +
            $"avg legal moves per board: {avgLegalMoves:F4}, " +
            $"random-swap match probability: {swapMatchProbability:F4}");

        Assert.That(avgLegalMoves, Is.GreaterThan(1.0));
        Assert.That(avgLegalMoves, Is.LessThan(80.0));
        Assert.That(swapMatchProbability, Is.GreaterThan(0.0));
        Assert.That(swapMatchProbability, Is.LessThan(0.5));
    }

    private static List<Position> Neighbours(int row, int col) => new()
    {
        new Position(row, col + 1),
        new Position(row + 1, col),
    };
}